### DISCUSSION TITLE
Raspberry Non-Arduino example in CI

### DISCUSSION THREAD

**Discussion by jgromes** [jgromes/RadioLib#736] [General]
@Mesteery since the Raspberry Pi example was added, it would be nice to have it also added to the CI. Would it be possible for you to share your current configuration? It would be very useful for me as a starting point.

**Comment by Mesteery** [comment]
I currently use xmake to build radiolib and other dependencies.
Here is the configuration I used to compile the example (xmake.lua):
```lua
add_rules("mode.debug", "mode.release")

package("pigpio")
    set_homepage("https://github.com/joan2937/pigpio")
    set_description("pigpio is a C library for the Raspberry which allows control of the General Purpose Input Outputs (GPIO).")
    set_license("Unlicense")

    add_urls("https://github.com/joan2937/pigpio/archive/refs/tags/$(version).tar.gz",
             "https://github.com/joan2937/pigpio.git")
    add_versions("79", "c33738a320a3e28824af7807edafda440952c05d")

    add_deps("cmake")
    add_syslinks("pthread")

    on_install(function (package)
        local configs = {}
        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
        import("package.tools.cmake").install(package, configs)
    end)

package("RadioLib")
    set_homepage("https://github.com/jgromes/RadioLib/")
    set_description("Universal wireless communication library for embedded devices ")
    set_license("MIT")

    add_urls("https://github.com/jgromes/RadioLib/archive/refs/tags/$(version).tar.gz",
             "https://github.com/jgromes/RadioLib.git")
    add_versions("5.7.0", "ead59223faaf225087b1db8da852b49d5e5947d9")

    on_install(function (package)
        local configs = {}
        io.writefile("xmake.lua", [[
            add_rules("mode.release", "mode.debug")
            add_defines("RADIOLIB_DEBUG")
            add_defines("RADIOLIB_VERBOSE")
            target("RadioLib")
               set_kind("static")
               add_files("src/**.cpp")
               add_includedirs("src")
               add_headerfiles("src/(**.h)")
        ]])
        import("package.tools.xmake").install(package, configs)
    end)
package_end()

add_requires("RadioLib master")
add_requires("pigpio develop")

target("radio")
    set_kind("binary")
    add_includedirs("src")
    add_packages("RadioLib", "pigpio")
    add_files("src/*.cpp")
```
(It's not very clean, it was only for testing)
I installed deps with `xmake q`
Compiled the project with `xmake`
and then I run it with `xmake r`, but for me pigio asked sudo, so I used `sudo` with the path to bin (`xmake r` show it at the end: `error: execv(...binpath...)`)
___

The ideal would be to add `xmake.lua` in the root of this project, which would support xmake (if we decide to use this toolchain for non-arduino) with something like that:
```lua
set_project("RadioLib")
set_version("6.0.0")
set_warnings("all", "error")

if is_mode("debug") then
    add_defines("RADIOLIB_DEBUG")
end

add_includedirs("src")

target("RadioLib")
  set_kind("static")
  add_files("src/(**.cpp)")
  add_headerfiles("src/(**.h)")
```
and so a project using radiolib would have for example this config
```lua
add_rules("mode.debug", "mode.release")

package("pigpio")
    set_homepage("https://github.com/joan2937/pigpio")
    set_description("pigpio is a C library for the Raspberry which allows control of the General Purpose Input Outputs (GPIO).")
    set_license("Unlicense")

    add_urls("https://github.com/joan2937/pigpio/archive/refs/tags/$(version).tar.gz",
             "https://github.com/joan2937/pigpio.git")
    add_versions("79", "c33738a320a3e28824af7807edafda440952c05d")

    add_deps("cmake")
    add_syslinks("pthread")

    on_install(function (package)
        local configs = {}
        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
        import("package.tools.cmake").install(package, configs)
    end)

package("RadioLib")
    set_homepage("https://github.com/jgromes/RadioLib/")
    set_description("Universal wireless communication library for embedded devices ")
    set_license("MIT")

    add_urls("https://github.com/jgromes/RadioLib/archive/refs/tags/$(version).tar.gz",
             "https://github.com/jgromes/RadioLib.git")
    add_versions("5.7.0", "ead59223faaf225087b1db8da852b49d5e5947d9")

    on_install(function (package)
        import("package.tools.xmake").install(package)
    end)
package_end()

add_requires("RadioLib master")
add_requires("pigpio develop")

target("radio")
    set_kind("binary")
    add_includedirs("src")
    add_packages("RadioLib", "pigpio")
    add_files("src/*.cpp")
```
I'm sure there are errors/misses and things that can be improved in both configs, as I'm fairly new to xmake.

**Reply by jgromes** [reply]
Thanks! I will probably stick to CMake if at all possible, since it's a lot more widespread (and by default included in most current Linux distributions).

I'll see if I can setup a self-hosted GitHub runner on a Raspberry Pi to have it compile (and maybe do some tests?) from the CI.

**Reply by jgromes** [reply]
@Mesteery I modified the Raspberry pi example to only use Cmake, feel free to check it out - IMHO it's a lot simpler than the xmake config ;) Tested on Raspberry pi and SX1302 LoRaWAN hat.

No luck with the runner yet, unfortuantely.

**Reply by Mesteery** [reply]
Awesome! I’m happy that it worked. I think there is a way to make a workflow for it by emulating arm or using cross compilation, I will try.
Also pigpio doesn’t seem to require rpi specific header. We could also in last resort just compile it for the available runner’s arch, and if it compile successfully , it will probably also compile successfully for Pi.

**Reply by jgromes** [reply]
I'd rather have it compile on real Rpi, since it opens up some interesting possibilities, like hardware-in-the loop testing.

**Reply by Mesteery** [reply]
Yes, of course, it's better but it's still a plus while waiting for the hardware/runner installation.

**Reply by Mesteery** [reply]
Also with the cmake config, how can I enable debug and verbose?

**Reply by jgromes** [reply]
It's shown in the cmake file in the RPi example, the command is like this (assuming PROJECT_NAME is storing the name of the target):

```cmake
target_compile_definitions(${PROJECT_NAME} 
  PUBLIC RADIOLIB_DEBUG 
  PUBLIC RADIOLIB_VERBOSE)  
```

**Reply by Mesteery** [reply]
When I compile my project with xmake, it does not find the radiolib header (xmake support cmake libs), however it find pigpio header without issue (which also use cmake)

**Reply by Mesteery** [reply]
Okay, I success to make it work, with the following config:
```cmake
cmake_minimum_required(VERSION 3.13)

project(radiolib)

file(GLOB_RECURSE RADIOLIB_SOURCES
  "src/*.cpp"
)

add_library(RadioLib ${RADIOLIB_SOURCES})

target_include_directories(RadioLib
    PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>
           $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)

include(GNUInstallDirs)

install(TARGETS RadioLib
  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
)

install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/src/
  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
  FILES_MATCHING PATTERN "*.h"
)
```
I don't know anything about cmake, but I've been looking at the config of a lot of libs using cmake for a while now, and this worked (btw all of them use install() but there is not install in the config you created).

**Reply by jgromes** [reply]
`install()` is needed in RadioLib cmake file since your xmake config is calling install(). I guess that too is a valid way to use the library, I will have to investigate this a bit more.

**Reply by Mesteery** [reply]
Okay I was not clear. This is the config from RadioLib cmake. I patched RadioLib root CMakeList.txt to this config, to make it work. So I propose this as CMakeList for radiolib cmake.

**Reply by jgromes** [reply]
I understood that (sorry for not being clear enough myself ...) What I meant is that install is needed in root RadioLib CMake because your xmake file is calling install(). That's not the only way to include a library though. In the Raspberry pi non-Arduino example, the library is added as a subdirectory using the `add_subdirectory` command, hence no install is needed. I don't know which is more standard and will have to investigate a bit more.

**Reply by Mesteery** [reply]
> sorry for not being clear enough myself

It was totally clear, it's just my fault, I read cmake instead of xmake and it confused me, sorry.

From what I have seen, all cmake libs use install().

**Reply by jgromes** [reply]
Did a bit of digging, turns out there's a bit of thruth from both sides. It is true that `install` is commonly done by libraries using cmake. However, it's *shared* libraries, which is not a very common use case for a library that is meant to be used for programs running on microcontrollers - which is what I am mostly familiar with.

Still, it is a valid use case - especially on Raspberry Pi, it's actually quite handy to install the library using the common `cmake` + `sudo make install` commands, and have the library availalbe system-wide.

I updated the root CMake to reflect this, and also the self-hosted RPi runner, so now the example is part of CI too!

**Comment by Mesteery** [comment]
I noticed this on the last CI run
```
2023-05-07 07:41:53 gpioSetMode: pigpio uninitialised, call gpioInitialise()
[SX1261] Test:begin() = -2
2023-05-07 07:41:53 gpioWrite: pigpio uninitialised, call gpioInitialise()
```

**Reply by jgromes** [reply]
The -2 error is easily explainable - the LoRaWAN hat revision I have doesn't actually expose the SX1262 SPI, so I ordered a new version which should include this.

I think the errors from pigpio are probably from some race condition when the program is ending.

**Reply by jgromes** [reply]
@Mesteery finally got the new hat, and it's working! This is all enabled by your efforts, so thanks again.

https://github.com/jgromes/RadioLib/actions/runs/5017756634/jobs/8996309145

Well almost working, there's nothing on the SDR (I'm guessing there's an RF switch somewhere along the path) and the BUSY pin seems unconnected.

**Reply by Mesteery** [reply]
It's amazing! Thanks for your work!

**Comment by jp-bennett** [comment]
Hey @jgromes and @Mesteery I have an SX1302 bolted on to a Raspberry Pi 400. Do I understand those last two comments to indicate that this isn't fully working with radiolib? I'm trying to integrate this into Meshtastic, and I'm currently seeing transmits just stall and never actually send.

**Reply by jgromes** [reply]
Building and running on RPi is fully implemented. However, RadioLib doesn't support SX1302.

**Reply by jp-bennett** [reply]
> However, RadioLib doesn't support SX1302.
Ah. That would explain it. It seems to be close. It can see the SX1262 and even read the version string. Any hope of getting this working in the near term?

**Reply by jp-bennett** [reply]
Eh, scratch that. There's apparently an SX1261 in there, but intended for receive only.

**Reply by jgromes** [reply]
Wait, you said you have an SX1302. Did you mean you have some LoraWAN gateway module like the one? https://www.waveshare.com/sx1302-868m-lorawan-gateway-b.htm

Ther is an SX126x in there, but it's not intended to be used without the SX1302 concentrator.

**Reply by jp-bennett** [reply]
Basically exactly that. It's the waveshare SX1302 in 915. I saw the example with the same combination and thought it should work.

**Reply by jgromes** [reply]
OK, I see. The status on my side did not change, I was able to get it to "transmit" (the RPi finds the SX126x and is able to configure it and command it to start transmitting, nothing hangs), but there was no actual RF output. The BUSY pin of the SX126x is not connected despite what the waveshare docs claim, so that might be what is happening to you.

There might be a way to get it working by e.g. soldering a wire to the busy pin and figuring out any possible RF switch controls, but I didn't dig that deep since I need the gateway for LoRaWAN testing, so I really don't want ot damage it.
